docs/notes/2.22.x.md

```diff
@@ -66,6 +66,12 @@ The deprecation for the `platforms` field for the `pex_binary` and `pex_binaries
 
 Setting [the `orphan_files_behaviour = "ignore"` option](https://www.pantsbuild.org/2.22/reference/subsystems/yamllint#orphan_files_behavior) the `pants.backend.experimental.tools.yamllint` backend is now properly silent. It previously showed spurious warnings.
 
+#### Golang
+
+Pants will now automatically install a Golang toolchain when needed, removing the need for installing it externally. This can be configured by `[go-toolchain].known_versions`, `[go-toolchain].enabled`, and `[go-toolchain].version`. As of shipping, you can select a version from *1.18.10*, *1.19.13*, *1.20.13*, and *1.21.6*. Other versions can be configured by adding to the known versions. Note that versions after Go 1.21 are currently not official supported due to the coverage system redesign. PRs are welcome here.
+
+The `[go-toolchain].enabled` flag is shipped and immediately deprecated, with the goal of only supporting a Pants-provided Go toolchain in the future.
```

Comment:
Is this in fact what users want? What if a user still wants to provide their own Go toolchain? Maybe we should not deprecate until we actually decide and intend to remove support for user-provided Go toolchains?

Reply:
From a "upgrade/process/release manager" perspective (I know nothing about Go), just reiterating that it might be good to at least have a release or two for it to bake and generally try to make upgrades easier. If we deprecate immediately and there's some bugs:

- it'll be annoying for users to stick with the old version (they'll get warnings they can't action, constantly)
- those bugs may need to be treated as higher severity by us (Pants maintainers) and need "immediate" attention, because they'll potentially be breaking existing code or making it hard for new users to onboard. This might then mean we have to delay releases to wait for fixes... and it's nicer for everyone to not have any of that panic.

One approach might be to: add the option without an explicit default. If it's not set:

- emit a deprecation warning (i.e. tell people that it must be set and that the default will change), and
- fall back to whichever values matches the current behaviour (to ease the upgrade path by not changing behaviour immediately).

Future releases can then continue the phasing out, e.g. a fairly conservative potential path might be:

| release | option | default | old behaviour | new behaviour |
|---|---|---|---|---|
| 2.22 (this PR) | added, deprecation when not set | old | not deprecated, unchanged | added |
| 2.23 | default changed | new | unchanged | unchanged, default |
| 2.24 | no deprecation when not set (just get new behaviour silently) | new | unchanged | unchanged |
| 2.25 | deprecated when set at all | new | deprecated | unchanged |
| 2.26 | waiting... | new | waiting... | unchanged |
| 2.27 | removed | new | removed | unchanged |


It looks like Go backend is experimental, so we can almost certainly do a more aggressive timeline than that... but I think we at least want to have:

- one release where users can opt-in to either the old behaviour or the new behaviour, and doesn't get warnings once they do the opt-in
- one release where the old behaviour exists but has deprecation warnings

What do you think?